src/transaction.js

```diff
@@ -651,34 +646,46 @@ function applyMultiSignatures (
   // Make a copy
   var txObj = JSON.parse(JSON.stringify(_txObj))
 
-  // TODO: make it stateless
-  // Fix signature order
-  // var rsFixed = redeemScript.slice(2)
-  // var pubKeys = []
-
-  // // 30 was chosen arbitrarily as the minimum length
-  // // of a pubkey is 33
-  // while (rsFixed.length > 30) {
-  //   // Convert pushdatalength from hex to int
-  //   // Extract public key
-  //   var pushDataLength = parseInt(rsFixed.slice(0, 2), 16).toString(10)
-  //   var pubkey = Buffer.from(rsFixed.slice(2), 'hex').slice(0, pushDataLength).toString('hex')
-  //   pubKeys = pubKeys.concat(pubkey)
-
-  //   rsFixed = rsFixed.slice(2 + pubkey.length)
-  // }
-
-  // var unmatched = JSON.parse(JSON.stringify(signatures))
-
-  // const signaturesFixed = pubKeys.map(pubKey => {
-  //   const keyPair = secp256k1.keyFromPublic(pubKey)
-
-  //   var match    
+  let pubkeys = []
+  let pubkeyLen
+  let pubkey
+  for (let pos = 2; pos < redeemScript.length - 2 * 2;) // skipping OP_M, OP_N and OP_CHECKMULTISIG code
+  {
+    pubkeyLen = parseInt(redeemScript.slice(pos, pos + 2), 16)
+    pos += 2
+    if (pubkeyLen == 0)
+    {
+      return "ERROR"
+    }
+    pubkey = redeemScript.slice(pos, pos + pubkeyLen * 2)
+    pos += pubkeyLen * 2
+    pubkeys.push(pubkey)
+  }
 
-  //   unmatched.some((sig, i) => {
-  //     if (!sig) return false      
-  //   })
-  // })
+  // recreate the message to sign
+  let signingTx = signatureForm(txObj, i, redeemScript, hashcode)
+  let _buf16 = Buffer.alloc(4)
+  _buf16.writeUInt16LE(hashcode, 0)
+  let signingTxHex = serializeTx(signingTx)
+  let signingTxWithHashcode = signingTxHex + _buf16.toString('hex') // Sha256 it twice, according to spec 
+  let msg = zcrypto.sha256x2(Buffer.from(signingTxWithHashcode, 'hex')) // Signing it
+
+  // find the association between pubkeys and signatures
+  let sortedSignatures = []
+  let pubkeysUncompressed = pubkeys.map((x) => secp256k1_.publicKeyConvert(Buffer.from(x, 'hex'), false))
+  for (let pkuId = 0; pkuId < pubkeysUncompressed.length; pkuId++)
+  {
+    for (let sId = 0; sId < signatures.length; sId++)
+    {
+      var verified = secp256k1.verify(msg, signatures[sId].slice(0, signatures[sId].length - 2), pubkeysUncompressed[pkuId]) // remove '01' at the end
+      if (verified)
+      {
+        sortedSignatures.push(signatures[sId])
+        signatures.splice(sId, 1)
+        break
+      }
+    }
```

Comment:
It might be easier to read using `.find`
```
var verifiedSignature = signatures.find(signature => secp256k1.verify(msg, signature.slice(0, signature.length - 2), pubkeysUncompressed[pkuId]) === true); // remove '01' at the end

if (verified) {
  sortedSigantures.push(verifiedSignature);
}
```

Reply:
Yes, I agree your proposal is more idiomatic. I've removed the inner `for` loop, preferred to use `findIndex` in place of `find` so to allow removing the verifying signature (thus making next iterations on `pubkeysUncompressed` faster).